### PR TITLE
Do something about stack overflows

### PR DESCRIPTION
**Pull request by mjambon**
* Print a warning when the stack size exceeds 7 MB.
* Raise a `Stack_overflow` exception when the stack size exceeds half of the user-specified memory limit (requires `ulimit -s unlimited` or equivalent).

See discussions and plan: https://github.com/returntocorp/semgrep/issues/3623

PR checklist:
- [x] documentation is up to date
- [x] changelog is up to date

### CODE REVIEW

File: semgrep-core/src/cli/Main.ml [hunk 1]
```diff
@@ -441,12 +454,31 @@ let run_with_memory_limit limit_mb f =
...
+      else if stack_bytes > stack_warning && not !stack_already_warned then (
+        logger#warning
+          "large stack size: %d bytes. If a crash follows, you should suspect \
+           a stack overflow. Make sure the maximum stack size is set to \
+           'unlimited' or to a value greater than %d bytes so as to obtain an \
+           exception rather than a segfault."
+          stack_bytes mem_limit;
+        stack_already_warned := true)
     in
     let alarm = Gc.create_alarm limit_memory in
```
Comment (IagoAbal) [inline semgrep-core/src/cli/Main.ml]: > create_alarm f will arrange for f to be called at the end of each major GC cycle, starting with the current cycle or the next one. A value of type alarm is returned that you can use to call delete_alarm.

I guess there is no warranty that the alarm will be called before the stack overflows?

Comment (IagoAbal) [reply semgrep-core/src/cli/Main.ml]: But it should work for the cases that we have experienced, I believe that previous to the stack overflow you could see increased allocation. Have you tested this?

**Comment by aryx** [comment]
You need to add Memory_limiit.ml

**Review by aryx — CHANGES_REQUESTED** [review]
missing file

**Comment by mjambon** [comment]
From what I can tell, simply using the GC alarm causes a 6x slowdown. We may have not seen it before because we'd not set a GC alarm with the default settings, but now we do.

I made another branch that is the same as `develop` branch except it always create the GC alarm: #3643. Let's see if this affects performance.

**Comment by aryx** [comment]
Ok, for once the semgrep benchmark regression test played its role! Glad it can at least catch those big regressions :)

**Comment by mjambon** [comment]
I think it should work without any performance degradation now.

Stack overflow prevention doesn't work very well because it requires a lot of allocation per function call for the GC alarm to run frequently enough (see the unit tests).

File: CHANGELOG.md [hunk 2]
```diff
@@ -10,6 +10,7 @@ This project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html
 ### Fixed
 - Go: Match import module paths correctly (#3484)
 - core: Do not crash when is not possible to compute range info
+- eliminate 6x slowdown when using the '--max-memory' option
```
Comment (IagoAbal) [inline CHANGELOG.md:29]: Nice!

**Comment by IagoAbal** [comment]
Fixing the slowdown for `-max_memory` was a very nice side-effect of this PR.

The stack overflow detection I am not sure if it's worth while. It seems to be an issue to be solved in the OCaml compiler, and apparently it only affects ARM. It only works under very specific conditions (this is not discussed in the code or in `-max_stack`...), it must trigger a major collection at the right time. Do we know if it actually works for the stack overflows that we experienced?

I think the right move would be to fix Semgrep so it doesn't get killed if anything happens when analyzing some specific file.

**Comment by mjambon** [comment]
Yeah, `-max_stack` doesn't work well - I had to put a lot of artificial allocation in the `grow_stack` function used in the unit tests otherwise it wouldn't work. I think it's misleading: we could still have segfaults (on those archs) due to stack overflow but not realize it because we expect `-max_stack` to work.

I'm going to do the following:
- remove `-max_stack` completely
- keep the other things this branch fixes (fix the perf issue with `Gc.compact()`, monitor heap+stack memory rather than just heap with `-max_memory`)
- always create a GC alarm so that all our tests and all production code run with it.

**Pull request merged**